File: android/src/main/java/sh/autograph/KeyExchange.kt
package sh.autograph

internal class KeyExchange {
    companion object {
        init {
            System.loadLibrary("autograph")
        }

        private external fun autographTranscriptSize(): Int

        private external fun autographKeyExchange(
            transcript: ByteArray,
            ourSignature: ByteArray,
            sendingKey: ByteArray,
            receivingKey: ByteArray,
            isInitiator: Boolean,
            ourIdentityKeyPair: ByteArray,
            ourSessionKeyPair: ByteArray,
            theirIdentityKey: ByteArray,
            theirSessionKey: ByteArray,
        ): Boolean

        private external fun autographVerifyKeyExchange(
            transcript: ByteArray,
            ourIdentityKeyPair: ByteArray,
            theirIdentityKey: ByteArray,
            theirSignature: ByteArray,
        ): Boolean

        fun createTranscript(): ByteArray = ByteArray(autographTranscriptSize())

        fun keyExchange(
            isInitiator: Boolean,
            ourIdentityKeyPair: ByteArray,
            ourSessionKeyPair: ByteArray,
            theirIdentityKey: ByteArray,
            theirSessionKey: ByteArray,
        ): Array<ByteArray> {
            var transcript = createTranscript()
            var ourSignature = Cert.createSignature()
            var sendingKey = Message.createSecretKey()
            var receivingKey = Message.createSecretKey()
            val success =
                autographKeyExchange(
                    transcript,
                    ourSignature,
                    sendingKey,
                    receivingKey,
                    isInitiator,
                    ourIdentityKeyPair,
                    ourSessionKeyPair,
                    theirIdentityKey,
                    theirSessionKey,
                )
            if (!success) {
                throw RuntimeException("Key exchange failed")
            }
            return arrayOf(transcript, ourSignature, sendingKey, receivingKey)
        }

        fun verifyKeyExchange(
            transcript: ByteArray,
            ourIdentityKeyPair: ByteArray,
            theirIdentityKey: ByteArray,
            theirSignature: ByteArray,
        ) {
            val verified = autographVerifyKeyExchange(transcript, ourIdentityKeyPair, theirIdentityKey, theirSignature)
            if (!verified) {
                throw RuntimeException("Key exchange failed")
            }
        }
    }
}

public fun keyExchange(
    isInitiator: Boolean,
    ourIdentityKeyPair: ByteArray,
    ourSessionKeyPair: ByteArray,
    theirIdentityKey: ByteArray,
    theirSessionKey: ByteArray,
): Array<ByteArray> {
    return KeyExchange.keyExchange(isInitiator, ourIdentityKeyPair, ourSessionKeyPair, theirIdentityKey, theirSessionKey)
}

public fun verifyKeyExchange(
    transcript: ByteArray,
    ourIdentityKeyPair: ByteArray,
    theirIdentityKey: ByteArray,
    theirSignature: ByteArray,
) {
    return KeyExchange.verifyKeyExchange(transcript, ourIdentityKeyPair, theirIdentityKey, theirSignature)
}

File: apple/Sources/Autograph/KeyExchange.swift
import Clibautograph
import Foundation

func createTranscript() -> [UInt8] {
    createBytes(autograph_transcript_size())
}

public func keyExchange(
    _ isInitiator: Bool,
    _ ourIdentityKeyPair: [UInt8],
    _ ourSessionKeyPair: [UInt8],
    _ theirIdentityKey: [UInt8],
    _ theirSessionKey: [UInt8]
) throws -> ([UInt8], [UInt8], [UInt8], [UInt8]) {
    var transcript = createTranscript()
    var ourSignature = createSignature()
    var sendingKey = createSecretKey()
    var receivingKey = createSecretKey()
    let success = autograph_key_exchange(
        &transcript,
        &ourSignature,
        &sendingKey,
        &receivingKey,
        isInitiator,
        ourIdentityKeyPair,
        ourSessionKeyPair,
        theirIdentityKey,
        theirSessionKey
    )
    if !success {
        throw AutographError.keyExchange
    }
    return (transcript, ourSignature, sendingKey, receivingKey)
}

public func verifyKeyExchange(
    _ transcript: [UInt8],
    _ ourIdentityKeyPair: [UInt8],
    _ theirIdentityKey: [UInt8],
    _ theirSignature: [UInt8]
) throws {
    let verified = autograph_verify_key_exchange(
        transcript,
        ourIdentityKeyPair,
        theirIdentityKey,
        theirSignature
    )
    if !verified {
        throw AutographError.keyExchange
    }
}
